### PR TITLE
Stardew Valley: Fix Daggerfish, Cropsanity; Move Some Rules to Content Packs; Add Missing Shipsanity Location

### PR DESCRIPTION
**Pull request by Witchybun**
This update does the following:

- Fixes a rule on Daggerfish.  Daggerfish are caught in the pond at Highlands, which is behind logs which requires axe.  So its a logic bug that says you can catch them if you can just reach the general area.
- While Jourmaie did move a fair deal of code to content packs, there were bits missing which caused an issue where any cropsanity crops were not included as locations upon generation for mods.
- While I was at it, and studying how Content Packs with with Jouramie, I moved a fair bit of code to them.  Its not complete, but will be in time.  Some cases were because there isn't a full/clean method to do so yet.

### CODE REVIEW

File: worlds/stardew_valley/content/mods/archeology.py
Comment (Jouramie) [inline worlds/stardew_valley/content/mods/archeology.py:23]: Since `all_artifacts` is a list, to find whether an item is part of it or not, behind the scene it will iterate through the whole list each time, which is not really fast. 

It would be better to move it to its own for loop on the artifact directly, so we don't have to check if the item is an artifact or not. It could also be done with a generator in the same line when adding the sources with something like: 
```python
content.source_item(ModTrash.rusty_scrap, *(MachineSource(item=artifact, machine=ModMachine.grinder) for artifact in all_artifacts))
```

**Review by agilbert1412 — APPROVED** [review]
Looks good to me, but my approval is condition to also getting @Jouramie 's approval, as he is more familiar and skilled than me at that aspect of the code. I give him last say on this

File: worlds/stardew_valley/content/mods/archeology.py
Comment (Jouramie) [inline worlds/stardew_valley/content/mods/archeology.py:32]: Is the `str` needed here? I might be missing something, but I feel like it's not...

File: worlds/stardew_valley/content/mods/sve.py
Comment (Jouramie) [inline worlds/stardew_valley/content/mods/sve.py:69]: I think those statements don't do anything. My understanding is that `source_item` add sources to an item, but in that case no source is given, so it does nothing.

If the intention is to remove the items, the best solution would be to pop them from `content.game_items`.

**Review by Jouramie — CHANGES_REQUESTED** [review]
Mostly look good, just 3 little comments

File: worlds/stardew_valley/content/mods/archeology.py
Comment (Witchybun) [reply worlds/stardew_valley/content/mods/archeology.py:32]: I used a str call here as when pulled the item has a tag of "any".  Its known to be a string but it just looked bad in the IDE lmao

File: worlds/stardew_valley/content/mods/archeology.py
Comment (Witchybun) [reply worlds/stardew_valley/content/mods/archeology.py:23]: Believe I've done this.  Also split off the iteration and made it a bit different.  I won't resolve this yet so you know it can be looked over again.

**Review by Jouramie — APPROVED** [review]
Feedback was addressed, LGTM

**Comment by black-sliver** [comment]
Can someone point me to the fix? I only see refactoring - next time, please split refactor and fix into two separate PRs that can be looked at individually.

**Comment by Jouramie** [comment]
@black-sliver The fix is in worlds/stardew_valley/data/fish_data.py, the region of the Daggerfish has been changed to Highlands Pond. This fix also required to create the region (in region_names.py, rules.py, etc), which didn't exist before.

The fix for cropsanity is literally the refactoring. In the 6.x.x PR, we introduced a new way to handle the content of the different mods. With this change, the location for cropsanity are based on what Content Packs are active, and what they content. The issue with cropsanity was that vanilla crops data was moved to content packs, but not the crops from SVE and Distant Lands. So refactoring SVE and Distant Lands fixed the issue.

**Comment by Witchybun** [comment]
The refactor IS the fix, as Jouramie pointed out.  I had to align with current content patcher features because it is the only way seeds are added unless I just made horrible code to step all over it.

**Comment by black-sliver** [comment]
> The refactor IS the fix

> While I was at it, and studying how Content Packs with with Jouramie, I moved a fair bit of code to them. Its not complete, but will be in time. Some cases were because there isn't a full/clean method to do so yet.

This sounds contradictory.

Anyway, so far I got

* the use of `__post_init__` is a code smell - I'm kinda fine with the asserts, assuming there is a reason for them, however

```python
    seasons: Tuple[str, ...] = Season.all
[...]
        if isinstance(self.seasons, str):
            super().__setattr__("seasons", (self.seasons,))
```
If seasons was a Tuple[str, ...], as the typing claims, then the isinstance would not do anything. If the seasons is a string, the typing is wrong.

* SVEFish.dulse_seaweed seems to have different logic? Or did I miss something?

* I don't really understand the separation. The new `highlands_pond = "Highlands Pond"` is in `strings/region_names.py`, the fish in `data//fish_data.py`, but the Location in `content/mods/sve.py`. This makes absolutely no sense to me. Either there is a mod providing things, or there isn't.

**Comment by agilbert1412** [comment]
> * I don't really understand the separation. The new `highlands_pond = "Highlands Pond"` is in `strings/region_names.py`, the fish in `data//fish_data.py`, but the Location in `content/mods/sve.py`. This makes absolutely no sense to me. Either there is a mod providing things, or there isn't.

I can't speak for the dulse seaweed or tuple stuff, but as for the content packs stuff, the short answer is that content packs are a work in progress thing. The long term vision is to move everything to content packs, for maximum modularity, but the current state does not have everything in content packs. I think less than half, although that would be hard to calculate.

Last I checked, fish aren't in content packs yet, so they are still all together in `data//fish_data.py`. They will eventually move to content packs, like crops already did, for example.

This might sound difficult to believe, considering the size of past PRs, but we are actually **trying** not to submit too big changes at a time, so Jouramie build content packs in "steps", that can be merged and are functional independently, specifically so he wouldn't need to move everything to content packs all at once. We established a timeline during the development of 6.x.x, where I was making features, and he was making content packs, and we tries to finish around the same time, so that we could submit a 6.x.x PR will all features, and a reasonable amount of content packing, without holding either back because of the other.

This also, kinda, explains the "inconsistency". Basically, Jouramie moved crops (among other things) to content packs, while Albrekka (WitchyBun) was working on the SVE support for 1.6 (and other mods, many had changes that needed the apworld to know about), so she did not follow very closely the specifics of Jouramie's work. This lead to her not realizing she needed to move some things (or, more specifically, she wrote some things in parallel to the move, in the "old style"), and this caused the cropsanity bug mentioned above. The new cropsanity relies on content packs entirely, but the crops weren't in them (heavily simplified, as I wasn't part of that, but you get the rough idea). So the refactoring being seen in this PR is Albrekka noticing this, and moving the *modded* content that should be in content packs, to content packs. Any modded equivalent to vanilla content that Jouramie had already moved, basically.

And while not all of it is specifically this bug, it is likely that it is causing other bugs, or at the very least strongly increasing the risk of other bugs, by being inconsistent, and therefore not being automatically caught by things like content pack unit tests, for example. So this is here saying "I was supposed to do content pack crops, and now cropsanity is broken. To fix cropsanity, I need content pack crops, and I was also supposed to do content packs [X,Y,Z], so I should do it too"

**Comment by Witchybun** [comment]
Dulse Seaweed was a fish in old SVE.  Current SVE has moved it to be equivalent to "Trash", or similar to other algae.  Its no longer a fish by SVE's definition, so its not going to be in the fish logic package.

I'm not sure I understand the concern regarding the regions string comment.  We've used the region_names.py for all string names regardless of mod.

**Comment by black-sliver** [comment]
Does that mean seaweed is now available every season, or does that mean sea weed doesn't exist anymore, in which case, why does it still exist, but with different logic?

> I'm not sure I understand the concern regarding the regions string comment.

It seems inconsistent to design a system that separates mods, but then only use it for a small subset of things.

However agilbert already answered that part.

**Comment by Witchybun** [comment]
You can catch it in most seasons but not every season.  Its just also not a fish, by the very definition of the game.  Its dropped as its no longer a fish, and not considered for Fishsanity, because its not a fish.

**Comment by Witchybun** [comment]
For the record, this means that:
- In Stardew Valley's code, its not in the fish category.  Its in a similar category to algae.
- When you hook a Dulse Seaweed, before you had to fight for it in a minigame; you now catch it immediately.

It has all the hallmarks of not being a fish so I removed it from the location list.

**Comment by black-sliver** [comment]
I am referring to this line: https://github.com/ArchipelagoMW/Archipelago/pull/3626/files#diff-f913646060d5b27266ef38a0f09738e2f2659fea2f2533e8970b612e983982ddR141

**Comment by black-sliver** [comment]
> I believe that was what the question was about?

Yes, but the text above opens another question:

> It has all the hallmarks of not being a fish so I removed it from the location list.

It is still in the "list of fish names" though. That is `SVEFish.dulse_seaweed` rather than `SVEWaterItem.dulse_seaweed` or similar, adding to the confusion.

**Comment by Witchybun** [comment]
Does the recent push resolve the issue for you?

**Comment by Witchybun** [comment]
A note.  Some players have noted they can't seem to find this modded book.  I also have had trouble.  Makes me think it doesn't exist.  I don't want to push something players can't get, so, while I dislike adding more to this, its minor; it removes Digging Like Worms from the multiworld.

**Comment by Jouramie** [comment]
Personally, I would only have removed the source from the content pack because the item and location will probably come back eventually, but I don't mind. LGTM

**Review by Jouramie — APPROVED** [review]

**Comment by black-sliver** [comment]
The first thing [in my first post](https://github.com/ArchipelagoMW/Archipelago/pull/3626#issuecomment-2229543539) wasn't replied to yet

**Comment by Witchybun** [comment]
Should be resolved now.  Was something Jouramie offered to me early on.  We can walk it back and do something better later.

File: worlds/stardew_valley/strings/fish_names.py
Comment (black-sliver) [inline worlds/stardew_valley/strings/fish_names.py:150]: Don't want to be super annoying, but If things should get split into mods, why not put the same SVE prefix as used in all other SVE groups?
```suggestion
class SVEWaterItem:
    dulse_seaweed = "Dulse Seaweed"
```

File: worlds/stardew_valley/content/mods/sve.py
Comment (black-sliver) [inline worlds/stardew_valley/content/mods/sve.py:13]:
```suggestion
from ...strings.fish_names import WaterItem, SVEFish, SVEWaterItem
```

File: worlds/stardew_valley/content/mods/sve.py
Comment (black-sliver) [inline worlds/stardew_valley/content/mods/sve.py:141]:
```suggestion
        SVEWaterItem.dulse_seaweed: (ForagingSource(regions=(Region.beach,), other_requirements=(FishingRequirement(Region.beach),)),),
```

File: worlds/stardew_valley/data/recipe_data.py
Comment (black-sliver) [inline worlds/stardew_valley/data/recipe_data.py:8]:
```suggestion
from ..strings.fish_names import Fish, SVEFish, WaterItem, DistantLandsFish, SVEWaterItem

```

File: worlds/stardew_valley/data/recipe_data.py
Comment (black-sliver) [inline worlds/stardew_valley/data/recipe_data.py:198]:
```suggestion
seaweed_salad = shop_recipe(SVEMeal.seaweed_salad, Region.fish_shop, 1250, {SVEWaterItem.dulse_seaweed: 2, WaterItem.seaweed: 2, Ingredient.oil: 1}, ModNames.sve)

```

**Comment by black-sliver** [comment]
If the biggest picture is to separate mods (into content packs), then the size of the collection is not what matters, but the fact that new stuff should follow the new style.

**Comment by Witchybun** [comment]
I don't really agree but if it gets this merged I will nod my head about it.

**Comment by black-sliver** [comment]
> I don't really agree

agilbert wrote:

> The long term vision is to move everything to content packs, for maximum modularity, but the current state does not have everything in content packs.

Maybe I am missing something, but to me that reads as "new code should be split per mod" so that the long term vision can be reached at some point.

> Move background names around or something idk

Renaming ModTrash to ArchaeologyTrash is just the middle finger here, right? You change code that has nothing to do with your PR.

Please add that change to the first post of your PR and the PR title to officially make this feature creep
or revert that change.

> Add missing shipsanity location

This should probably also add a "* add missed location" in the top post of the PR.

**Comment by Witchybun** [comment]
Its not a middle finger.  If I didn't change it, you would have noted it, since its a similar situation and would've delayed the PR.

**Comment by black-sliver** [comment]
>  If I didn't change it, you would have noted it

I would not, otherwise I would have already.

> since its a similar situation

The difference between "Dulse Seaweed" and "Scrap Rust" is that the collection holding "Scrap Rust" already existed and only the code referencing it moved, while "Dulse Seaweed" created a new collection that I think would've gone against what I believe to be the goal.

> I'm not him.

That doesn't matter. https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/CODEOWNERS#L167 clearly lists agilbert as the world maintainer, so I will stick to their long term vision when looking at code that gets added to their world.

This may sound unbelievable, but I am not picky because I want to be annoying or block this PR, but I want to make this and future code changes as good as possible so we don't end up with a hot mess that needs unnecessarily big refactors later on.

Also please keep in mind that it does not help if you communicated or talked about something somewhere else if you don't link to it and the reviewer didn't accidentally see it. I have no idea what kind of discussion led to the initial (or current) version of this PR.
For example, from the original message it was not clear what was or wasn't refactor and/or fix, because the 3rd bullet does not mention fix, but looked like it would cover 99% of this PR. I think a good messages would be something like

> Move SVE and DistandLands to a Content Pack so Cropsanity can pick it up, fixing an issue where Cropsanity would not include them.

I'm also still unsure what the change to ArchaeologyContentPack actually does other than refactor, since you did not mention this in your initial message nor does it define any crops.

**Review by agilbert1412 — APPROVED** [review]

**Comment by Witchybun** [comment]
I'll take that to heart.  Give me a bit and I'll clean this up a bit.  The archaeology thing, I can revert.  Can be fixed later since its not relevant now.

> I'm also still unsure what the change to ArchaeologyContentPack actually does other than refactor, since you did not mention this in your initial message nor does it define any crops.

Its essentially for the same reason I gave before.  When fixing cropsanity, I had to move a lot of code to the content packs.  It feels unfinished to not move everything which is possible to move, as its not logically consistent to do one thing and not another.  Its just what I do.  There's a few example Kaito can probably attest to with the core mod where I've done that; he'll offer a suggestion, I'll see its logical conclusion, and I'll just do all of it that I recall.

**Comment by Witchybun** [comment]
I've been told I don't have to merge every time the branch is out of dead :skull:

**Review by black-sliver — APPROVED** [review]

**Comment by black-sliver** [comment]
Yep, when we merge the thing, it gets applied "on top" anyway.

You need to "update branch" when there is either a conflict (two commits/PRs changing the same file) or there is reason to believe a different change could impact your code (such as a new test or change to a file you import from failing for your code).

Code looks good now, I think.

What I missed so far is that removal of "Digging Like Worms" was not mentioned in this PR and the commit also has no explanatory message, so if you could give the reason here, that'd be amazing.

**Comment by Witchybun** [comment]
Digging Like Worms was an item listed in Archaeology's files and seemed gettable, but during a beta no player was capable of getting it, which makes me wonder if it was included as actually gettable at all.

My general design philosophy on adding modded content is to try to keep in the spirit of the mod author best I can with some alterations for the sake of a randomizer.  If the book isn't gettable, the dev didn't intend for it to be gettable, meaning I shouldn't give it to the player.  So removing it just puts the design philosophy back to where it should be.

**Pull request merged**
